substrate/frame/staking/src/lib.rs

```diff
@@ -1223,3 +1223,59 @@ impl BenchmarkingConfig for TestBenchmarkingConfig {
 	type MaxValidators = frame_support::traits::ConstU32<100>;
 	type MaxNominators = frame_support::traits::ConstU32<100>;
 }
+
+/// Controls validator disabling
+pub trait DisablingStrategy<T: Config> {
+	/// Make a decision if an offender should be disabled or not. The result is a `Vec` of validator
+	/// indices that should be disabled
+	fn make_disabling_decision(
+		offender_stash: &T::AccountId,
+		slash_era: EraIndex,
+		currently_disabled: &Vec<u32>,
+	) -> Vec<u32>;
+}
+
+/// Implementation of [`DisablingStrategy`] which disables validators from the active set up to a
+/// threshold. `DISABLING_THRESHOLD_FACTOR` is the factor of the maximum disabled validators in the
+/// active set. E.g. setting this value to `3` means no more than 1/3 of the validators in the
+/// active set can be disabled in an era.
+/// By default a factor of 3 is used which is the byzantine threshold.
+pub struct UpToThresholdDisablingStrategy<const DISABLING_THRESHOLD_FACTOR: usize = 3>;
+
+impl<const DISABLING_THRESHOLD_FACTOR: usize>
+	UpToThresholdDisablingStrategy<DISABLING_THRESHOLD_FACTOR>
+{
+	/// Disabling limit calculated from the total number of validators in the active set. When
+	/// reached no more validators will be disabled.
+	pub fn disable_threshold(validators_len: usize) -> usize {
+		validators_len.saturating_sub(1) / DISABLING_THRESHOLD_FACTOR
```

Comment:
> We could consider using checked div here and return an error if it fails.

You are worried about division by zero here or there is something more I am missing?

If just former - I'm 99% sure that setting `DISABLING_THRESHOLD_FACTOR` to zero will yield a compilation error (the computation happens compile time). I don't mind adding an integrity test but isn't it an overkill to check a constant with it?

Reply:
That's a good catch. I was sure setting DISABLING_THRESHOLD_FACTOR to 0 will yield a compilation error but this is not the case.

I think such errors should be reported at compile time so I'll add an integrity test for it.

Reply:
On a second thought this check can't be done as an integrity test because we can't assume runtime uses `UpToThresholdDisablingStrategy`. At the same time testing `UpToThresholdDisablingStrategy` on it's own doesn't work because the value is unknown until the struct is instantiated.

The only workaround I can think about is something like a 'static_assert' which is a bit hacky for const parameters but still [can be done](https://github.com/nvzqz/static-assertions/issues/40#issuecomment-1458897730).